src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp

```diff
@@ -1047,3 +1052,47 @@ void ControllerScriptInterfaceLegacy::softStart(int deck, bool activate, double
     // activate the ramping in scratchProcess()
     m_ramp[deck] = true;
 }
+
+QByteArray ControllerScriptInterfaceLegacy::convertCharset(
+        const ControllerScriptInterfaceLegacy::WellKnownCharsets targetCharset,
+        const QString& value) {
+    switch (targetCharset) {
+    case WellKnownCharsets::Latin1:
+    case WellKnownCharsets::ISO_8859_1:
+        return convertCharset(QStringLiteral("ISO-8859-1"), value);
+    case WellKnownCharsets::Latin9:
+    case WellKnownCharsets::ISO_8859_15:
+        return convertCharset(QStringLiteral("ISO-8859-15"), value);
+    case WellKnownCharsets::UCS2:
+    case WellKnownCharsets::ISO_10646_UCS_2:
+        return convertCharset(QStringLiteral("ISO-10646-UCS-2"), value);
+    default:
+        m_pScriptEngineLegacy->logOrThrowError(QStringLiteral("Unknown charset specified"));
+        return QByteArray();
+    }
+}
+
+QByteArray ControllerScriptInterfaceLegacy::convertCharset(
+        const QString& targetCharset, const QString& value) {
+#if QT_VERSION < QT_VERSION_CHECK(6, 4, 0)
+    auto* pCodec = QTextCodec::codecForName(targetCharset.toUtf8());
+    if (!pCodec) {
+        m_pScriptEngineLegacy->logOrThrowError(QStringLiteral("Unable to open encoder"));
+        return QByteArray();
+    }
+    return pCodec->makeEncoder(QTextCodec::Flag::ConvertInvalidToNull)->fromUnicode(value);
+#else
+#if QT_VERSION > QT_VERSION_CHECK(6, 8, 0)
+    const auto* const encoderName = targetCharset.data();
+#else
+    auto* const encoderName = targetCharset.toUtf8().data();
```

Comment:
```suggestion
    QByteArray encoderNameArray = targetCharset.toUtf8();
    const char* encoderName = encoderNameArray.constData();
```
This line worked for me only in normal operation, but not if I run it in the debugger. In the debugger .toUtf8() returned only garbage. I think the lifetime of the array wasn't guaranteeed here. After introducing the deep copy into a local array it works also in the debugger.
I only tested this branch (Windows 11 with Qt6.5) but the others look similar.

Reply:
Won't line 1086 have the same problem?

Reply:
Yes, that is what I meant with the others look similar.

Reply:
K. Met pe fix that.

Reply:
Wait… Why was it useful to remove `.toUtf8()` for Qt>=6.8.0 again? Can't I just do:

```c++
QStringEncoder(targetCharset.targetCharset.toUtf8().data()
```

Reply:
> Wait… Why was it useful to remove .toUtf8() for Qt>=6.8.0 again? Can't I just do:

Because it does a utf16 to utf8 conversion along with a heap allocation at runtime while on Qt6.8 [QStringEncoder can take a `QAnyStringView`](https://doc.qt.io/qt-6/qstringencoder.html#QStringEncoder-1) which eliminates that. So the removal of the `toUtf8` call was an optimization opportunity thats only possible on new enough Qt versions.

Reply:
Regardless, Jörg is correct. `.toUtf8().data()` creates a dangling reference so we need to store that (unnecessary) copy separately.

Reply:
Ok, I wrote a slightly different version in my last commit. Is it OK ?

Reply:
The last commit does not address the issue. You need a deep copy for this string reference - I'm not sure if this is needed for the StringView in Qt6.8 as well, but I guess so.


``` C++
QByteArray ControllerScriptInterfaceLegacy::convertCharset(
        const QString& targetCharset, const QString& value) {
    QByteArray encoderNameArray = targetCharset.toUtf8();
#if QT_VERSION < QT_VERSION_CHECK(6, 4, 0)
    auto* pCodec = QTextCodec::codecForName(encoderNameArray);
    if (!pCodec) {
        m_pScriptEngineLegacy->logOrThrowError(QStringLiteral("Unable to open encoder"));
        return QByteArray();
    }
    return pCodec->makeEncoder(QTextCodec::Flag::ConvertInvalidToNull)->fromUnicode(value);
#else
#if QT_VERSION > QT_VERSION_CHECK(6, 8, 0)
    QAnyStringView encoderName = QAnyStringView::fromArray(encoderNameArray);
#else
    const char* encoderName = encoderNameArray.constData();
#endif
    QStringEncoder fromUtf16 = QStringEncoder(
            encoderName, QStringEncoder::Flag::ConvertInvalidToNull);
    if (!fromUtf16.isValid()) {
        m_pScriptEngineLegacy->logOrThrowError(QStringLiteral("Unable to open encoder"));
        return QByteArray();
    }
    return fromUtf16(value);
#endif
}
```

I only tested this with Qt6.5

Reply:
Hmm… Ok I won't pretend I understand what you are both saying here. I amended my last commit with @JoergAtGithub's version. I tested running the tests in debug. They are passing here for Qt6.8. I just changed `QT_VERSION > QT_VERSION_CHECK(6, 8, 0)` to `QT_VERSION >= QT_VERSION_CHECK(6, 8, 0)`.

Reply:
> I'm not sure if this is needed for the StringView in Qt6.8 as well, but I guess so.

Its not, which is the primary difference between `QLatin1String` and `QAnyStringView` and why the `QAnyStringView` is more efficient in this case. Its essentially a `variant<string_view, u8string_view, u16string_view>` so you can efficiently construct it from most common string formats.